ccw.core/src/clj/ccw/leiningen/classpath_container.clj

```diff
@@ -353,50 +364,102 @@
+        project-name (e/project-name java-project)
+        error-fn (fn [[workspace-job exception]]
+                   (let [[jresource message] (resource-message exception java-project)
+                         error-message (format "Leiningen Managed Dependencies issue: %s" message)]
+                     ;; TODO enhance this in the future ... (more accurate problem markers)
+                     (report-container-error jresource error-message exception)
+                     error-message))
+        finally-fn (fn [workspace-job] (doto ;; we refresh the target folder outside the workspace-root lock
+                             (e/workspace-job
+                              (format "Refreshing project %s" project-name)
+                              (fn [^IProgressMonitor monitor]
+                                (refresh-target-folder java-project monitor)
+                                Status/OK_STATUS))
+                           (.setUser true)
+                           (.schedule)))
+        job (e/cancelable-workspace-job
+             (format "Update project dependencies for project %s" project-name)
+             (fn [^IProgressMonitor monitor]
+               (letfn [(cancel! [] (e/throw-op-canceled! (format "The update of %s was canceled" project-name) monitor))]
+                 (let [_ (cancel!)
+                       lein-project (u/lein-project java-project :enhance-fn #(do (t/trace :leiningen %) (dissoc % :hooks)))
+                       deps (get-project-dependencies cancel! (.getName (e/project java-project)) lein-project)]
+                   (t/trace :leiningen "Project dependencies calculated, setting up the workspace...")
+                   ;; Here, get-project-dependencies has succeeded or thrown an error
+                   ;; it can take a long time, so we do not put it inside the workspace job which blocks on the workspace root
+                   (cancel!)
+                   (doto
+                       (e/workspace-job
+                        (format "Upgrade project build path for project %s" (e/project-name java-project))
+                        (fn [^IProgressMonitor monitor]
+                          (set-lein-container java-project deps)
+                          (delete-container-markers java-project)
+                          (save-project-dependencies java-project deps)
+                          Status/OK_STATUS))
+                     (.setUser true)
+                     ;; this rule is OK because we know the job needs it and will not take too long
+                     (.setRule (e/workspace-root))
+                     (.schedule))))
+               (if-not (.isCanceled monitor)
+                 Status/OK_STATUS
+                 Status/CANCEL_STATUS))
+             :error-handler! error-fn
+             :finally-handler! finally-fn)
+        new-state (swap! state start-updating project-id job)]
+    (when (can-update? new-state project-id)
```

Comment:
I don't think we should just be happy with "whoever wins". Unless there's a bug, if `update-dependencies` is called twice, it's because there has been changes to be taken into account. Meaning that the current job should stop as far as possible, to let the second job work with (presumably) more recent data.